File: app/src/main/java/com/android/captureinterfacexposed/ui/activity/SelectAppActivity.java
package com.android.captureinterfacexposed.ui.activity;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.os.AsyncTask;
import android.os.IBinder;
import android.text.TextUtils;
import android.text.format.Formatter;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.BaseAdapter;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.PopupMenu;
import android.widget.SearchView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import com.android.captureinterfacexposed.application.DefaultApplication;
import com.android.captureinterfacexposed.R;
import com.android.captureinterfacexposed.databinding.ActivitySelectAppBinding;
import com.android.captureinterfacexposed.ui.activity.base.BaseActivity;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class SelectAppActivity extends BaseActivity<ActivitySelectAppBinding> {
    private ListView appListView;
    private SearchView searchView;
    private HashMap<Integer, Integer> posMap = new HashMap<>();
    private ProgressDialog loadingDialog;
    private static List<ApplicationInfo> appList;
    private static ArrayList<AppItem> appItemList;
    private AppListAdapter adapter;

    @Override
    public void onCreate() {

        // TitleThemes
        binding.includeTitleBarSecond.tvTitle.setText(getText(R.string.select_app));
        binding.includeTitleBarSecond.ivBackButton.setOnClickListener(v -> onBackPressed());
        binding.includeTitleBarSecond.ivMoreButton.setOnClickListener(v -> showPopupMenu(binding.includeTitleBarSecond.ivMoreButton));

        appListView = binding.appListView;
        searchView = binding.searchView;

        if(appList == null || appItemList == null){
            loadingDialog = ProgressDialog.show(SelectAppActivity.this, "数据加载中", "请稍后...", true, false);
            new LoadDataTask().execute();
        } else {
            adapter = new AppListAdapter(appItemList);
            appListView.setAdapter(adapter);
        }

        searchView.setSubmitButtonEnabled(false);
        searchView.onActionViewExpanded();

        // 添加搜索框的监听器
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                if (TextUtils.isEmpty(newText)) {
                    // 没有输入的时候就显示所有
                    setAdapter();
                } else {
                    List<AppItem> list = new ArrayList<>();
                    for (ApplicationInfo app: appList) {
                        String appName = app.loadLabel(getPackageManager()).toString();
                        String packageName = app.packageName;
                        String versionName = getVersionName(packageName);
                        String size = getSize(app);
                        Drawable appIcon = app.loadIcon(getPackageManager());
                        String item = appName + " (" + packageName + ")";
                        if (item.contains(newText)) {
                            int index = appList.indexOf(app);
                            AppItem appItem = new AppItem(item,appName,versionName,size,appIcon,index);
                            list.add(appItem);
                        }
                    }
                    ArrayList<AppItem> new_appItemList = new ArrayList<>(list);
                    adapter = new AppListAdapter(new_appItemList);
                    appListView.setAdapter(adapter);
                }
                return true;
            }
        });

        appListView.setOnItemClickListener((parent, view, position, id) -> {
            String appName = appList.get(posMap.get(position)).loadLabel(SelectAppActivity.this.getPackageManager()).toString();
            String packageName = appList.get(posMap.get(position)).packageName;

            // 将选择的应用名称和包名返回给 MainActivity
            Intent intent = new Intent();
            intent.putExtra("app_name", appName);
            intent.putExtra("package_name", packageName);
            setResult(RESULT_OK, intent);
            finish();
        });
    }


    private class LoadDataTask extends AsyncTask<Void, Void, Void> {
        @Override
        protected Void doInBackground(Void... voids) {
            processData(); // 拉取数据
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            setAdapter(); // 刷新列表
            loadingDialog.dismiss(); // 关闭进度条
        }
    }

    public void processData() {
        // 获取本机已安装应用信息
        final List<ApplicationInfo> appList_ = getPackageManager().getInstalledApplications(PackageManager.GET_META_DATA);
        appList = new ArrayList<>();
        for (ApplicationInfo app : appList_){
            // 筛选有活动入口的应用程序
            if(DefaultApplication.entryActivityClassName(app.packageName) != null && !app.packageName.equals(getApplicationInfo().packageName)) {
                appList.add(app);
            }
        }
    }

    private void setAdapter(){
        appItemList = new ArrayList<>();
        for (ApplicationInfo app: appList) {
            String appName = app.loadLabel(getPackageManager()).toString();
            String packageName = app.packageName;
            String versionName = getVersionName(packageName);
            String size = getSize(app);
            Drawable appIcon = app.loadIcon(getPackageManager());
            int index = appList.indexOf(app);
            AppItem appItem = new AppItem(appName + " (" + packageName + ")",appName,versionName,size,appIcon,index);
            appItemList.add(appItem);
        }
        adapter = new AppListAdapter(appItemList);
        appListView.setAdapter(adapter);
    }

    private class AppListAdapter extends BaseAdapter {

        private final ArrayList<AppItem> appItemList;

        public AppListAdapter(ArrayList<AppItem> appItemList) {
            this.appItemList = appItemList;
        }

        @Override
        public int getCount() {
            return appItemList.size();
        }

        @Override
        public String getItem(int position) {
            return appItemList.get(position).toString();
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            if (convertView == null) {
                convertView = LayoutInflater.from(getApplicationContext()).inflate(R.layout.app_list_item, parent, false);
            }
            ImageView appIconImageView = convertView.findViewById(R.id.app_icon_image_view);
            TextView appNamesTextView = convertView.findViewById(R.id.app_name_text_view);
            TextView appVersionTextView = convertView.findViewById(R.id.app_version_text_view);
            TextView appSizeTextView = convertView.findViewById(R.id.app_size_text_view);
            int index = appItemList.get(position).index;
            Drawable appIcon = appItemList.get(position).appIcon;
            String appNames = appItemList.get(position).appNames;
            String versionName = appItemList.get(position).appVersion;
            String size = appItemList.get(position).appSize;

            appIconImageView.setImageDrawable(appIcon);
            appNamesTextView.setText(appNames);
            appVersionTextView.setText("版本: " + versionName);
            appSizeTextView.setText(size);
            appNamesTextView.setTextColor(getResources().getColor(R.color.firstTextColor));
            appVersionTextView.setTextColor(getResources().getColor(R.color.secondTextColor));
            appSizeTextView.setTextColor(getResources().getColor(R.color.secondTextColor));

            posMap.put(position, index);
            return convertView;
        }
    }
    private String getVersionName(String packageName) {
        PackageManager packageManager = getPackageManager();
        String versionName = "";
        try {
            PackageInfo packageInfo = packageManager.getPackageInfo(packageName, 0);
            versionName = packageInfo.versionName;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return versionName;
    }

    private String getSize(ApplicationInfo app) {
        String apkPath = app.sourceDir;
        long size = new File(apkPath).length();
        return Formatter.formatFileSize(this, size);
    }

    private static class AppItem {
        private final String appNames;
        private final String appName;
        private final String appVersion;
        private final String appSize;
        private final Drawable appIcon;
        private final int index;

        AppItem(String appNames, String appName, String appVersion, String appSize, Drawable appIcon, int index){
            this.appNames = appNames;
            this.appName = appName;
            this.appVersion = appVersion;
            this.appSize = appSize;
            this.appIcon = appIcon;
            this.index = index;
        }

        @NonNull
        @Override
        public String toString() {
            return "appNames: " + appNames +
                    "\nappName: " + appName +
                    "\nappVersion: " + appVersion +
                    "\nappSize" + appSize;
        }
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {
        if (ev.getAction() == MotionEvent.ACTION_DOWN) {
            View view = getCurrentFocus();
            if (isHideInput(view, ev)) {
                HideSoftInput(view.getWindowToken());
                view.clearFocus();
            }
        }
        return super.dispatchTouchEvent(ev);
    }
    // 判定是否需要隐藏
    private boolean isHideInput(View v, MotionEvent ev) {
        if (v instanceof EditText) {
            int[] l = {0, 0};
            v.getLocationInWindow(l);
            int left = l[0], top = l[1], bottom = top + v.getHeight(), right = left + v.getWidth();
            return !(ev.getX() > left) || !(ev.getX() < right) || !(ev.getY() > top) || !(ev.getY() < bottom);
        }
        return false;
    }
    // 隐藏软键盘
    private void HideSoftInput(IBinder token) {
        if (token != null) {
            InputMethodManager manager = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
            manager.hideSoftInputFromWindow(token, InputMethodManager.HIDE_NOT_ALWAYS);
        }
    }
    private void showPopupMenu(View anchorView) {
        PopupMenu popupMenu = new PopupMenu(this, anchorView);
        popupMenu.getMenuInflater().inflate(R.menu.menu_popup, popupMenu.getMenu());
        popupMenu.getMenu().findItem(R.id.restart_sys).setVisible(false);
        popupMenu.getMenu().findItem(R.id.restart_app).setVisible(false);
        popupMenu.setOnMenuItemClickListener(item -> {
            if (item.getItemId() == R.id.refresh_list) {
                searchView.setQuery("",false);
                appListView.setAdapter(null);
                appList.clear();
                appItemList.clear();
                loadingDialog = ProgressDialog.show(SelectAppActivity.this, "数据加载中", "请稍后...", true, false);
                new LoadDataTask().execute();
            }
            return true;
        });
        popupMenu.show();
    }
}

File: app/src/main/java/com/android/captureinterfacexposed/ui/activity/SelectAppActivity.kt
package com.android.captureinterfacexposed.ui.activity

import android.annotation.SuppressLint
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.graphics.drawable.Drawable
import android.os.IBinder
import android.text.TextUtils
import android.text.format.Formatter
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.BaseAdapter
import android.widget.EditText
import android.widget.ImageView
import android.widget.PopupMenu
import android.widget.SearchView
import android.widget.TextView
import androidx.activity.OnBackPressedCallback
import androidx.lifecycle.lifecycleScope
import com.android.captureinterfacexposed.R
import com.android.captureinterfacexposed.application.DefaultApplication.Companion.entryActivityClassName
import com.android.captureinterfacexposed.databinding.ActivitySelectAppBinding
import com.android.captureinterfacexposed.ui.activity.base.BaseActivity
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

class SelectAppActivity : BaseActivity<ActivitySelectAppBinding>() {
    companion object {
        private var appList: MutableList<ApplicationInfo>? = null
        private var appItemList: ArrayList<AppItem>? = null
    }
    private val posMap = HashMap<Int, Int>()
    private var adapter: AppListAdapter? = null
    private var isProcessBarStatus = false

    override fun onCreate() {
        binding.includeTitleBarSecond.tvTitle.text = getText(R.string.select_app)
        binding.includeTitleBarSecond.ivBackButton.setOnClickListener { finish() }
        binding.includeTitleBarSecond.ivMoreButton.setOnClickListener { showPopupMenu(binding.includeTitleBarSecond.ivMoreButton) }
        if (appList == null || appItemList == null) {
            lifecycleScope.launch { loadData() }
        } else {
            adapter = AppListAdapter(appItemList!!)
            binding.appListView.adapter = adapter
        }
        binding.searchView.isSubmitButtonEnabled = false
        binding.searchView.onActionViewExpanded()
        binding.searchView.clearFocus()

        // 添加搜索框的监听器
        binding.searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                if(isProcessBarStatus) return true
                if (TextUtils.isEmpty(newText)) {
                    // 没有输入的时候就显示所有
                    setAdapter()
                } else {
                    val list: MutableList<AppItem> = ArrayList()
                    for (app in appList!!) {
                        val appName = app.loadLabel(packageManager).toString()
                        val packageName = app.packageName
                        val versionName = getVersionName(packageName)
                        val size = getSize(app)
                        val appIcon = app.loadIcon(packageManager)
                        val item = "$appName ($packageName)"
                        if (item.contains(newText)) {
                            val index = appList!!.indexOf(app)
                            val appItem = AppItem(item, appName, versionName, size, appIcon, index)
                            list.add(appItem)
                        }
                    }
                    val newAppItemList = ArrayList(list)
                    adapter = AppListAdapter(newAppItemList)
                    binding.appListView.adapter = adapter
                }
                return true
            }
        })
        binding.appListView.onItemClickListener =
            OnItemClickListener { _: AdapterView<*>?, _: View?, position: Int, _: Long ->
                val appName = appList!![posMap[position]!!].loadLabel(this@SelectAppActivity.packageManager).toString()
                val packageName = appList!![posMap[position]!!].packageName
                // 将选择的应用名称和包名返回给 MainActivity
                val intent = Intent()
                intent.putExtra("app_name", appName)
                intent.putExtra("package_name", packageName)
                setResult(RESULT_OK, intent)
                finish()
            }
        onBackPressedDispatcher.addCallback(
            this, // lifecycle owner
            object : OnBackPressedCallback(true) {
                override fun handleOnBackPressed() {
                    if (!isProcessBarStatus) { finish() }
                }
            })
    }

    @SuppressLint("QueryPermissionsNeeded")
    private fun processData() {
        // 获取本机已安装应用信息
        val appList1 = packageManager.getInstalledApplications(PackageManager.GET_META_DATA)
        appList = ArrayList()
        for (app in appList1) {
            // 筛选有活动入口的应用程序
            if (entryActivityClassName(app.packageName) != null && app.packageName != applicationInfo.packageName) {
                (appList as ArrayList<ApplicationInfo>).add(app)
            }
        }
    }

    private fun setAdapter() {
        appItemList = ArrayList()
        for (app in appList!!) {
            val appName = app.loadLabel(packageManager).toString()
            val packageName = app.packageName
            val versionName = getVersionName(packageName)
            val size = getSize(app)
            val appIcon = app.loadIcon(packageManager)
            val index = appList!!.indexOf(app)
            val appItem = AppItem(
                "$appName ($packageName)", appName, versionName, size, appIcon, index
            )
            appItemList!!.add(appItem)
        }
        adapter = AppListAdapter(appItemList!!)
        binding.appListView.adapter = adapter
    }

    private inner class AppListAdapter(private val appItemList: ArrayList<AppItem>) :
        BaseAdapter() {
        override fun getCount(): Int {
            return appItemList.size
        }

        override fun getItem(position: Int): String {
            return appItemList[position].toString()
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View? {
            var view = convertView
            val holder: ViewHolder
            if (view == null) {
                view = LayoutInflater.from(applicationContext).inflate(R.layout.app_list_item, parent, false)
                holder = ViewHolder(view)
                view.tag = holder
            } else {
                holder = view.tag as ViewHolder
            }
            val index = appItemList[position].index
            val appIcon = appItemList[position].appIcon
            val appNames = appItemList[position].appNames
            val versionName = appItemList[position].appVersion
            val size = appItemList[position].appSize
            holder.appIconImageView.setImageDrawable(appIcon)
            holder.appNamesTextView.text = appNames
            holder.appVersionTextView.text = resources.getString(R.string.app_version, versionName)
            holder.appSizeTextView.text = size
            holder.appNamesTextView.setTextColor(resources.getColor(R.color.firstTextColor))
            holder.appVersionTextView.setTextColor(resources.getColor(R.color.secondTextColor))
            holder.appSizeTextView.setTextColor(resources.getColor(R.color.secondTextColor))
            posMap[position] = index
            return view
        }

        private inner class ViewHolder(view: View) {
            val appIconImageView: ImageView = view.findViewById(R.id.app_icon_image_view)
            val appNamesTextView:TextView = view.findViewById(R.id.app_name_text_view)
            val appVersionTextView:TextView = view.findViewById(R.id.app_version_text_view)
            val appSizeTextView:TextView = view.findViewById(R.id.app_size_text_view)
        }
    }

    private fun getVersionName(packageName: String): String {
        val packageManager = packageManager
        var versionName = ""
        try {
            val packageInfo = packageManager.getPackageInfo(packageName, 0)
            versionName = packageInfo.versionName
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
        return versionName
    }

    private fun getSize(app: ApplicationInfo): String {
        val apkPath = app.sourceDir
        val size = File(apkPath).length()
        return Formatter.formatFileSize(this, size)
    }

    private class AppItem(
        val appNames: String,
        val appName: String,
        val appVersion: String,
        val appSize: String,
        val appIcon: Drawable,
        val index: Int
    ) {
        override fun toString(): String {
            return """
                appNames: $appNames
                appName: $appName
                appVersion: $appVersion
                appSize$appSize
                """.trimIndent()
        }
    }

    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
        if (ev.action == MotionEvent.ACTION_DOWN) {
            val view = currentFocus
            if (isHideInput(view, ev)) {
                hideSoftInput(view!!.windowToken)
                view.clearFocus()
            }
        }
        return super.dispatchTouchEvent(ev)
    }

    // 判定是否需要隐藏
    private fun isHideInput(v: View?, ev: MotionEvent): Boolean {
        if (v is EditText) {
            val l = intArrayOf(0, 0)
            v.getLocationInWindow(l)
            val left = l[0]
            val top = l[1]
            val bottom = top + v.getHeight()
            val right = left + v.getWidth()
            return ev.x <= left || ev.x >= right || ev.y <= top || ev.y >= bottom
        }
        return false
    }

    // 隐藏软键盘
    private fun hideSoftInput(token: IBinder?) {
        if (token != null) {
            val manager = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
            manager.hideSoftInputFromWindow(token, InputMethodManager.HIDE_NOT_ALWAYS)
        }
    }

    private fun showPopupMenu(anchorView: View) {
        val popupMenu = PopupMenu(this, anchorView)
        popupMenu.menuInflater.inflate(R.menu.menu_popup, popupMenu.menu)
        popupMenu.menu.findItem(R.id.restart_sys).isVisible = false
        popupMenu.menu.findItem(R.id.restart_app).isVisible = false
        popupMenu.setOnMenuItemClickListener { item: MenuItem ->
            if (item.itemId == R.id.refresh_list) {
                binding.searchView.setQuery("", false)
                binding.appListView.adapter = null
                appList!!.clear()
                appItemList!!.clear()
                lifecycleScope.launch { loadData() }
            }
            true
        }
        popupMenu.show()
    }

    private suspend fun loadData() {
        inProcessBar()
        withContext(Dispatchers.IO) {
            processData()
        }
        outProcessBar()
        withContext(Dispatchers.Main) {
            setAdapter() // 刷新列表
        }
    }

    private fun inProcessBar(){
        isProcessBarStatus = true
        binding.progressBar.visibility = View.VISIBLE
        binding.includeTitleBarSecond.ivBackButton.isEnabled = false
        binding.includeTitleBarSecond.ivMoreButton.isEnabled = false
    }

    private fun outProcessBar(){
        isProcessBarStatus = false
        binding.progressBar.visibility = View.GONE
        binding.includeTitleBarSecond.ivBackButton.isEnabled = true
        binding.includeTitleBarSecond.ivMoreButton.isEnabled = true
    }

}